### PR TITLE
[WIP] Add Ovis2.5 

### PR DESCRIPTION
**Pull request by thisisiron**
# What does this PR do?

This PR adds support for Ovis2.5, a multimodal vision-language model designed for native-resolution visual perception and reflective reasoning.

Ovis2.5 is released in two variants:
- [Ovis2.5-9B](https://huggingface.co/AIDC-AI/Ovis2.5-9B)
- [Ovis2.5-2B](https://huggingface.co/AIDC-AI/Ovis2.5-2B)

This is still a draft PR. Further integration and testing to follow.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by Rocketknight1** [comment]
cc @zucchini-nlp

**Comment by zucchini-nlp** [comment]
Reviewing, sorry for delay

File: docs/source/en/model_doc/ovis2_5.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+-->
+
+# Ovis2.5
+
+## Overview
+
+The  is an updated version of the [Ovis](https://arxiv.org/abs/2405.20797) model developed by the AIDC-AI team at Alibaba International Digital Commerce Group. 
+
+[Ovis2.5](https://arxiv.org/abs/2508.11737) is a multimodal vision-language model, available in 2B and 9B parameter versions. It introduces a native-resolution vision transformer (NaViT) that processes images at their original resolutions, preserving both fine details and global layout—crucial for visually dense inputs like complex charts and documents. The model also features a novel reflective reasoning capability, including self-checking and revision, accessible via an optional thinking mode at inference time.
+
```
Comment (zucchini-nlp) [inline docs/source/en/model_doc/ovis2_5.md:24]: Lte's add an abstract in italics to follow general model doc guidelines

File: docs/source/en/model_doc/ovis2_5.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+from PIL import Image
+import requests
+import torch
+from torchvision import io
+from typing import Dict
+from transformers.image_utils import load_images, load_video
+from transformers import AutoModelForVision2Seq, AutoTokenizer, AutoProcessor
+
+model = AutoModelForVision2Seq.from_pretrained(
+    "",
```
Comment (zucchini-nlp) [inline docs/source/en/model_doc/ovis2_5.md:46]: missing model id?

File: docs/source/en/model_doc/ovis2_5.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+from PIL import Image
+import requests
+import torch
+from torchvision import io
+from typing import Dict
+from transformers.image_utils import load_images, load_video
+from transformers import AutoModelForVision2Seq, AutoTokenizer, AutoProcessor
+
+model = AutoModelForVision2Seq.from_pretrained(
```
Comment (zucchini-nlp) [inline docs/source/en/model_doc/ovis2_5.md:45]: `AutoModelForVision2Seq` is deprecated, we need to ise `AutoModelForImageTextToText`

File: docs/source/en/model_doc/ovis2_5.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+print(messages)
+
+inputs = processor(
+    images=[image],
+    text=messages,
+    return_tensors="pt",
+)
+inputs = inputs.to("cuda:0")
+inputs['pixel_values'] = inputs['pixel_values'].to(torch.bfloat16)
+
```
Comment (zucchini-nlp) [inline docs/source/en/model_doc/ovis2_5.md:72]: can do shorter with

`processor.apply_chat_template(messages, tokenize=True, return_dict=True, return_tensors='pt')`

See [llava docs](https://huggingface.co/docs/transformers/en/model_doc/llava#usage-examples) for an example

File: docs/source/en/model_doc/ovis2_5.md [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+with torch.inference_mode():
+    output_ids = model.generate(**inputs, max_new_tokens=128, do_sample=False)
+    generated_ids = [output_ids[len(input_ids):] for input_ids, output_ids in zip(inputs.input_ids, output_ids)]
+    output_text = processor.batch_decode(generated_ids, skip_special_tokens=True)
+    print(output_text)
+```
+
+<!-- TODO: Convert Ovis2 to Ovis2.5  -->
```
Comment (zucchini-nlp) [inline docs/source/en/model_doc/ovis2_5.md:80]: todo and delete comment

File: src/transformers/models/ovis2_5/modular_ovis2_5.py [hunk 6]
```diff
@@ -0,0 +1,420 @@
...
+        super().__init__(
+            vision_config=vision_config,
+            text_config=text_config,
+            image_token_id=image_token_id,
+            visual_indicator_token_ids=visual_indicator_token_ids,
+            vocab_size=vocab_size,
+            hidden_size=hidden_size,
+            **kwargs
+        )
+        # TODO: add video token id
```
Comment (zucchini-nlp) [inline src/transformers/models/ovis2_5/modular_ovis2_5.py:118]: todo?

File: src/transformers/models/ovis2_5/modular_ovis2_5.py [hunk 7]
```diff
@@ -0,0 +1,420 @@
...
+
+        # siglip2 naflex
+        if self.num_patches > 0:
+            self.patch_embedding = nn.Linear(
+                in_features=config.num_channels * self.patch_size * self.patch_size,
+                out_features=self.embed_dim,
+            )
+            if self.preserve_original_pe:  # TODO: rename
+                self.position_embedding_size = int(self.num_patches**0.5)
+        else:
```
Comment (zucchini-nlp) [inline src/transformers/models/ovis2_5/modular_ovis2_5.py:136]: if the released weights have only `num_patches == 0` or the other way, we can keep only one code path

File: src/transformers/models/ovis2_5/modular_ovis2_5.py [hunk 8]
```diff
@@ -0,0 +1,420 @@
...
+        # Apply patch embeddings to already patchified pixel values
+        target_dtype = self.patch_embedding.weight.dtype
+        if isinstance(self.patch_embedding, nn.Linear):
+            patch_embeds = self.patch_embedding(pixel_values.to(dtype=target_dtype))
+        elif isinstance(self.patch_embedding, nn.Conv2d):
+            pixel_values = pixel_values.view(
+                -1, self.config.num_channels * self.config.temporal_patch_size, self.patch_size, self.patch_size
+            )
+            patch_embeds = self.patch_embedding(pixel_values.to(dtype=target_dtype))
+            patch_embeds = patch_embeds.reshape(-1, self.embed_dim)
```
Comment (zucchini-nlp) [inline src/transformers/models/ovis2_5/modular_ovis2_5.py:168]: continuing on above comment, this looks weird if we are checking on module type. In case we need both position embedding types, i thing it's better to have separate classes for each and just call it

File: src/transformers/models/ovis2_5/modular_ovis2_5.py [hunk 9]
```diff
@@ -0,0 +1,420 @@
...
+                thw = t * h * w
+                pe = F.interpolate(positional_embeddings, size=(h, w), mode="bicubic", align_corners=False)
+                pe = pe.permute(0, 2, 3, 1).reshape(1, h * w, -1)
+                pe = pe[0].repeat(t, 1)
+                pe = pe.reshape(
+                    t, h // self.hidden_stride, self.hidden_stride, w // self.hidden_stride, self.hidden_stride, -1
+                )
+                pe = pe.permute(0, 1, 3, 2, 4, 5).reshape(thw, -1)
+                pos_embed_new[cnt : cnt + thw] = pe
+                cnt += thw
```
Comment (zucchini-nlp) [inline src/transformers/models/ovis2_5/modular_ovis2_5.py:191]: could be in `self.interpolate_position_embeddings(grid_twh)` to be aligned with other ViT models

File: src/transformers/models/ovis2_5/modular_ovis2_5.py [hunk 10]
```diff
@@ -0,0 +1,420 @@
...
+        self.rotary_pos_emb = Ovis2_5VisionRotaryEmbedding(config.hidden_size // config.num_attention_heads // 2)
+        self.patch_size = config.patch_size
+        self.hidden_stride = config.hidden_stride
+        self.window_size = config.window_size
+        self.spatial_merge_unit = config.hidden_stride * config.hidden_stride
+        self.fullatt_block_indexes = (
+            None if config.fullatt_block_indexes is None else [int(i) for i in config.fullatt_block_indexes.split("|")]
+        )
+
+    def rot_pos_emb(self, grid_thw):
```
Comment (zucchini-nlp) [inline src/transformers/models/ovis2_5/modular_ovis2_5.py:213]: looks like the rope from Qwen2.5-VL, maybe we should use modular with Qwen2.5-Vision then?

File: src/transformers/models/ovis2_5/modular_ovis2_5.py [hunk 11]
```diff
@@ -0,0 +1,420 @@
...
+                - 0 for tokens that are **masked**.
+                [What are attention masks?](../glossary#attention-mask)
+            output_attentions (`bool`, *optional*):
+                Whether or not to return the attentions tensors of all attention layers. See `attentions` under
+                returned tensors for more detail.
+            output_hidden_states (`bool`, *optional*):
+                Whether or not to return the hidden states of all layers. See `hidden_states` under returned tensors
+                for more detail.
+            return_dict (`bool`, *optional*):
+                Whether or not to return a [`~utils.ModelOutput`] instead of a plain tuple.
```
Comment (zucchini-nlp) [inline src/transformers/models/ovis2_5/modular_ovis2_5.py:309]: let's use `@autodocstring` instead to auto-generate docs

File: src/transformers/models/ovis2_5/modular_ovis2_5.py [hunk 12]
```diff
@@ -0,0 +1,420 @@
...
+            if self.fullatt_block_indexes is None or index in self.fullatt_block_indexes:
+                cu_seqlens_tmp = cu_seqlens
+            else:
+                cu_seqlens_tmp = cu_window_seqlens
+            if self.gradient_checkpointing and self.training:
+                hidden_states = self._gradient_checkpointing_func(
+                    block.__call__, hidden_states, cu_seqlens_tmp, position_embeddings
+                )
+            else:
+                hidden_states = block(hidden_states, cu_seqlens_tmp, position_embeddings)
```
Comment (zucchini-nlp) [inline src/transformers/models/ovis2_5/modular_ovis2_5.py:355]: GC can be handled better by using `GradientCheckpointingLayer` :)

**Review by zucchini-nlp — COMMENTED** [review]
Not sure if the PR is ready for review. Just left a few general comments

**Comment by thisisiron** [comment]
@zucchini-nlp, I’m still working on adding the model. Sorry for the confusion. I’ll ping you once it’s ready for review. Thank you for taking the time to review.

**Comment by zucchini-nlp** [comment]
Oke, i will unsubscribe myself then :)

**Comment by zucchini-nlp** [comment]
@thisisiron hey, is this PR still active? Ovis2.5 is getting community interest, so lmk if you need help with finalizing the PR

**Comment by thisisiron** [comment]
@zucchini-nlp, I’ve been quite busy recently and haven’t been able to make proper progress on this. Apologies for the delay. Please feel free to close the PR so that someone else can continue the work.

**Comment by zucchini-nlp** [comment]
No probs :) In case you want to take over and finish the PR, cc @mahimairaja. The next step would be to address comments from last review
